huaweicloud/services/iotda/resource_huaweicloud_iotda_dataforwarding_rule.go

```diff
@@ -0,0 +1,732 @@
+	}
+
+	if resp.RuleId == nil {
+		return diag.Errorf("error creating IoTDA data forwarding rule: id is not found in API response")
+	}
+
+	d.SetId(*resp.RuleId)
+	m := d.Get("targets").(*schema.Set)
+	// create action rule
+	targets, err := buildActionTargets(m.List(), d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	for _, v := range targets {
+		_, err := client.CreateRuleAction(&v)
+		if err != nil {
+			return diag.Errorf("error add targets to IoTDA data forwarding rule: %s", err)
+		}
+	}
+
+	// enable forwarding
+	if d.Get("enabled").(bool) {
+		_, err = client.UpdateRoutingRule(&model.UpdateRoutingRuleRequest{
+			RuleId: d.Id(),
+			Body: &model.UpdateRuleReq{
+				Active: utils.Bool(true),
+			},
+		})
+		if err != nil {
+			return diag.Errorf("error activing the IoTDA data forwarding rule: %s", err)
+		}
+	}
+
+	return ResourceDataForwardingRuleRead(ctx, d, meta)
+}
+
+func ResourceDataForwardingRuleRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	response, err := client.ShowRoutingRule(&model.ShowRoutingRuleRequest{RuleId: d.Id()})
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving IoTDA data forwarding rule")
+	}
+
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("name", response.RuleName),
+		d.Set("trigger", flattenTrigger(response.Subject)),
+		d.Set("description", response.Description),
+		d.Set("select", response.Select),
+		d.Set("where", response.Where),
+		d.Set("enabled", response.Active),
+		d.Set("space_id", response.AppId),
+		setTargetsToState(d, client, d.Id()),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func ResourceDataForwardingRuleUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	if d.HasChange("targets") {
+		o, n := d.GetChange("targets")
+		oldTargetSet := o.(*schema.Set)
+		newTargetSet := n.(*schema.Set)
+
+		// delete
+		for _, v := range oldTargetSet.Difference(newTargetSet).List() {
+			ruleAction := v.(map[string]interface{})
+			_, err = client.DeleteRuleAction(&model.DeleteRuleActionRequest{ActionId: ruleAction["id"].(string)})
+			if err != nil {
+				return diag.Errorf("error updating targets of IoTDA data forwarding rule: %s", err)
+			}
+		}
+
+		// add and update
+		for _, v := range newTargetSet.List() {
+			target := v.(map[string]interface{})
+			channel := target["type"].(string)
+			channelDetail, err := buildChannelDetail(target, channel)
+			if err != nil {
+				return diag.FromErr(err)
+			}
+
+			if id, ok := target["id"]; ok && len(id.(string)) > 0 {
```

Comment:
I think the  "ok" judgement is just enough

Reply:
no, OK always is true,  so add the length check